Feature/implement details request

adamconder opened the pull request:
- Implementing route and connector to request GRO API using child's forenames, surname and date of birth

Commit:
Refactored BirthConnector to call only one request method

Commit:
lookup() now uses partial funcitons to call connector method

Commit:
BirthConnectorSpec uses Payload for argument

Commit:
Add test scenario for unsucessful match on child details

Commit:
Refactor exceptions

Commit:
feature/implement-details-request: updated to form details url with encode.

adamconder committed:
adding comments and refactoring model

Commit:
feature/implement-details-request: updated record objects to fix compile error.

Commit:
feature/implement-details-request: updated to read array.

adamconder committed:
refactoring json parsin

Commit:
Add test scenario for match based on child details

Commit:
Merge changes

Commit:
feature/implement-details-request: updated test for details match true.

Commit:
Add multiple match test scenarios

adamconder committed:
refactor to do tail recursive matching

Commit:
Refactor Matching functionality

adamconder committed:
details-request: refactoring config

adamconder committed:
add log to warn when failed to parse

adamconder committed:
add splunk logging and grafana metrics

adamconder committed:
add logs to lookup service

adamconder commented on app/uk/gov/hmrc/brm/models/response/Record.scala:
```diff
@@ -14,20 +14,32 @@
+    )(Record.apply _)
 }
+
+//case class Records(records : List[Record])
```
Need to remove commented code

adamconder commented on app/uk/gov/hmrc/brm/services/LookupService.scala:
```diff
@@ -63,56 +64,96 @@ trait LookupService extends LookupServiceBinder {
+  private[LookupService] def referenceNumberIncludedPF(implicit hc: HeaderCarrier, payload: Payload): PartialFunction[Payload, Future[HttpResponse]] = {
+    case payload@Payload(Some(birthReferenceNumber), _, _, _, _) =>
+      /**
+        * TODO: Return a generic interface BirthResponse which can use Reads/Adapter to map JsValue to case class
```
Remove this comment block

adamconder commented on app/uk/gov/hmrc/brm/services/MatchingService.scala:
```diff
@@ -17,24 +17,36 @@
   val CLASS_NAME: String = this.getClass.getCanonicalName
 
-  def performMatch(input: Payload, response: GroResponse, matchingType: MatchingType.Value)(implicit hc: HeaderCarrier) : ResultMatch = {
+  val matchOnMultiple: Boolean
```
Make this protected[MatchingService]

adamconder commented on app/uk/gov/hmrc/brm/services/MatchingService.scala:
```diff
@@ -17,24 +17,36 @@
-    val result = algorithm.performMatch(input, response)
+    val result = algorithm.performMatch(input, records, matchOnMultiple)
+
+    if (result.isMatch) BRMAudit.logEventRecordFound(hc)
```
Remove this for the splunk auditing as this would not audit for this metric as the match would be false

adamconder commented on app/uk/gov/hmrc/brm/utils/CommonUtil.scala:
```diff
@@ -18,93 +18,62 @@ package uk.gov.hmrc.brm.utils
 
-trait ControllerUtil {
-  self: Controller =>
+trait ControllerUtil extends Controller with BrmException {
```
Change this to be a `BRMBaseController` which is extended from the MatchingController which itself extends HMRC BaseController

chris-ianson replied:
The BirthEventsController now extends the BRMBaseController (previously ControllerUtil) I've also split it into it's own file.

adamconder commented on test/uk/gov/hmrc/brm/connectors/BirthConnectorSpec.scala:
```diff
@@ -74,26 +76,23 @@ class BirthConnectorSpec extends UnitSpec with WithFakeApplication with MockitoS
     }
 
-
     "getChildDetails returns json response" in {
```
@chris-ianson Add unit tests for getDetails

adamconder commented on test/uk/gov/hmrc/brm/controllers/BirthEventsControllerSpec.scala:
```diff
@@ -109,9 +76,42 @@ class BirthEventsControllerSpec
-    "POST /birth-registration-matching-proxy/match" should {
+    "POST /birth-registration-matching-proxy/match NOT INCLUDING reference number" should {
+      "return JSON response of false on unsuccessful child detail match" in {
+        when(MockController.service.groConnector.getChildDetails(Matchers.any())(Matchers.any())).thenReturn(Future.successful(httpResponse(Json.parse("[]"))))
```
Add unit test to ensure handling exceptions etc

adamconder commented on test/uk/gov/hmrc/brm/models/GroResponseSpec.scala:
```diff
@@ -860,12 +861,12 @@ class GroResponseSpec extends UnitSpec {
 
-    "return GroResponse object with Status object when blockedRegistration key is excluded" in {
-      val result = jsonStatusFlagsExcludingBlockedRegistration.validate[GroResponse]
+    "return Record object with Status object when blockedRegistration key is excluded" in {
```
Add unit tests for List[Record]

chris-ianson replied:
Added a test for List[Record]

adamconder commented on test/uk/gov/hmrc/brm/services/MatchingServiceSpec.scala:
```diff
@@ -32,6 +32,10 @@ import uk.gov.hmrc.play.test.UnitSpec
 
 class MatchingServiceConfigSpec extends UnitSpec with MockitoSugar with BeforeAndAfterAll {
 
+  //  val attemptMatchOnMultipleResultConfig
```
remove comment

adamconder reviewed (commented):
Please make changes requested in comments

Commit:
Resolve pull-request issues

adamconder committed:
changing application-json-logger

- changing application-json-logger from DEBUG => WARN for production
** app-config ** is no longer respected in production

adamconder committed:
Merge branch 'feature/implement-details-request' of github.com:hmrc/birth-registration-matching into feature/implement-details-request

adamconder committed:
changing log level

adamconder committed:
changing and remove logs

adamconder committed:
changing loggers

adamconder commented on app/uk/gov/hmrc/brm/services/MatchingService.scala:
```diff
@@ -17,24 +17,34 @@
 import uk.gov.hmrc.play.http.HeaderCarrier
 
-trait MatchingService {
+trait MatchingService extends ServicesConfig {
```
@chris-ianson This no longer needs to extend ServicesConfig

adamconder committed:
fix log file name:

Commit:
Remove ServicesConfig

Commit:
Removed check on body length within BirthEventsControllerSpec

chris-ianson reviewed (commented):
The check on contentLength is returning ByteString causing the tests to fail. The content returned looked like the contentType. Not sure how the changes we did yesterday introduced this?

Commit:
Add NameFormat to detailsURI

Commit:
feature/implement-details-request: removed lowercase format.
